solr/modules/monitor/src/java/org/apache/solr/monitor/search/ReverseSearchComponent.java

```diff
@@ -0,0 +1,207 @@
+
+  private static final String SOLR_MONITOR_CACHE_NAME_KEY = "solrMonitorCacheName";
+  private static final String SOLR_MONITOR_CACHE_NAME_DEFAULT = "solrMonitorCache";
+  private String solrMonitorCacheName = SOLR_MONITOR_CACHE_NAME_DEFAULT;
+
+  private QueryDecomposer queryDecomposer;
+  private Presearcher presearcher;
+  private final SolrMatcherSinkFactory solrMatcherSinkFactory = new SolrMatcherSinkFactory();
+  private PresearcherFactory.PresearcherParameters presearcherParameters;
+
+  @Override
+  public void init(NamedList<?> args) {
+    super.init(args);
+    Object solrMonitorCacheName = args.remove(SOLR_MONITOR_CACHE_NAME_KEY);
+    if (solrMonitorCacheName != null) {
+      this.solrMonitorCacheName = (String) solrMonitorCacheName;
+    }
+    presearcherParameters = new PresearcherFactory.PresearcherParameters();
+    SolrPluginUtils.invokeSetters(presearcherParameters, args);
+  }
+
+  @Override
+  public void prepare(ResponseBuilder rb) {
+    var req = rb.req;
+    var documentBatch = documentBatch(req);
+    var matcherSink = solrMatcherSinkFactory.build(documentBatch, rb.req.getContext());
+    Query preFilterQuery = presearcher.buildQuery(documentBatch.get(), getTermAcceptor(rb.req));
+    List<Query> mutableFilters =
+        Optional.ofNullable(rb.getFilters()).map(ArrayList::new).orElseGet(ArrayList::new);
+    rb.setQuery(new MatchAllDocsQuery());
+    mutableFilters.add(preFilterQuery);
+    var searcher = req.getSearcher();
+    MonitorQueryCache solrMonitorCache =
+        (SharedMonitorCache) searcher.getCache(this.solrMonitorCacheName);
+    SolrMonitorQueryDecoder queryDecoder = new SolrMonitorQueryDecoder(req.getCore());
+    mutableFilters.add(
+        new MonitorPostFilter(
+            new SolrMonitorQueryCollector.CollectorContext(
+                solrMonitorCache, queryDecoder, matcherSink)));
+    rb.setFilters(mutableFilters);
+  }
+
+  @SuppressWarnings({"unchecked"})
+  private DocumentBatchVisitor documentBatch(SolrQueryRequest req) {
+    Object jsonParams = req.getJSON().get("params");
+    if (!(jsonParams instanceof Map)) {
+      throw new SolrException(SolrException.ErrorCode.BAD_REQUEST, "need params");
+    }
+    var paramMap = (Map<?, ?>) jsonParams;
+    var documents = paramMap.get(MONITOR_DOCUMENTS_KEY);
+    if (!(documents instanceof List)) {
+      throw new SolrException(SolrException.ErrorCode.BAD_REQUEST, "need documents list");
+    }
+    List<Document> luceneDocs = new ArrayList<>();
+    for (var document : (List<?>) documents) {
+      if (!(document instanceof Map)
+          || !((Map<?, ?>) document).keySet().stream().allMatch(key -> key instanceof String)) {
+        throw new SolrException(
+            SolrException.ErrorCode.BAD_REQUEST, "document needs to be a string-keyed map");
+      }
+      var docAsMap = (Map<Object, Object>) document;
+      docAsMap.putIfAbsent(
+          req.getSchema().getUniqueKeyField().getName(), UUID.randomUUID().toString());
```

Comment:
Wondering if it would be simpler to require the caller to supply the key field.

Reply:
I added this based on user feedback. Alerts are typically not triggered on the Id field yet the id field is the only guaranteed required field. It seemed unnecessary to force users to supply an id field just to satisfy the validation of a different use-case (i.e. they aren't indexing this document but rather "alerting" on its contents so they shouldn't  be required to have _any_ fields). Obviously this doesn't help with required non-id fields but that's a more specific use-case. Because the fix was simple enough, I figured I'd add it. Is there a downside to this I'm not considering?

Reply:
I guess I should correct myself.. id/uniqueKey field isn't _strictly_ required but I think most schemas will have it (and most likely _should_ have it) and so I wanted to accomodate it...

Reply:
My "would be simpler" thoughts were something along these lines:
 * "the key field must always be supplied" is clear to the user (and for the code)
 * "the key field must never be supplied" is clear to the user (and for the code)
 * "if the key field is not (fully) supplied it will be auto-generated" raises a bunch of questions:
   * will the auto-generation influence the results, hopefully not the results directly but perhaps results ordering?
   * what if the auto-generation (for missing key fields) clashes with supplied key field values?
   * can the auto-generation support my particular type of key field type e.g. composite routing -- https://solr.apache.org/guide/solr/latest/deployment-guide/solrcloud-shards-indexing.html#document-routing
   * in a sharded setup, could different shards auto-generate the same key (or could a key auto-generated on one shard clash with a supplied key on another shard), unlikely in practice though?

Perhaps a middle ground could be _"the key field must be supplied for all documents or for none of the documents, and if it's not supplied then ... will be used to auto-generate and that will only work for ... fields, if you use another key field type then you must supply the keys"_ or something like that.